File: frontend/src/lib/stores/events.ts
/**
 * Events/Webhooks store for managing webhook subscriptions and event messages
 */

import { create } from 'zustand';
import { apiClient } from '../api';

/**
 * Event message type based on Svix MessageOut
 */
export interface EventMessage {
  id: string;
  eventType: string;
  payload: Record<string, unknown>;
  timestamp: string;
  channels?: string[];
}

/**
 * Subscription type based on Svix EndpointOut
 */
export interface Subscription {
  id: string;
  url: string;
  channels?: string[];
  createdAt: string;
  updatedAt: string;
  description?: string;
  disabled?: boolean;
}

/**
 * Message attempt type based on Svix MessageAttemptOut
 */
export interface MessageAttempt {
  id: string;
  url: string;
  msgId: string;
  endpointId: string;
  response: string;
  responseStatusCode: number;
  timestamp: string;
  status: number;
  triggerType: number;
}

/**
 * Subscription with message attempts response type
 */
export interface SubscriptionWithAttempts {
  endpoint: Subscription;
  message_attempts: MessageAttempt[];
}

/**
 * Create subscription request type
 */
export interface CreateSubscriptionRequest {
  url: string;
  event_types: string[];
  secret?: string;
}

/**
 * Update subscription request type
 */
export interface UpdateSubscriptionRequest {
  url?: string;
  event_types?: string[];
}

/**
 * Subscription secret response type
 */
export interface SubscriptionSecret {
  key: string;
}

interface EventsState {
  subscriptions: Subscription[];
  messages: EventMessage[];
  isLoadingSubscriptions: boolean;
  isLoadingMessages: boolean;
  error: string | null;

  // Actions
  fetchSubscriptions: () => Promise<void>;
  fetchMessages: (eventTypes?: string[]) => Promise<void>;
  fetchSubscription: (subscriptionId: string) => Promise<SubscriptionWithAttempts>;
  createSubscription: (request: CreateSubscriptionRequest) => Promise<Subscription>;
  updateSubscription: (subscriptionId: string, request: UpdateSubscriptionRequest) => Promise<Subscription>;
  deleteSubscription: (subscriptionId: string) => Promise<void>;
  deleteSubscriptions: (subscriptionIds: string[]) => Promise<void>;
  fetchSubscriptionSecret: (subscriptionId: string) => Promise<SubscriptionSecret>;
  clearEvents: () => void;
}

export const useEventsStore = create<EventsState>((set, get) => ({
  subscriptions: [],
  messages: [],
  isLoadingSubscriptions: false,
  isLoadingMessages: false,
  error: null,

  fetchSubscriptions: async () => {
    set({ isLoadingSubscriptions: true, error: null });
    try {
      const response = await apiClient.get('/events/subscriptions');
      if (!response.ok) {
        throw new Error(`Failed to fetch subscriptions: ${response.status}`);
      }
      const subscriptions = await response.json();
      set({ subscriptions, isLoadingSubscriptions: false });
    } catch (error) {
      set({
        error: error instanceof Error ? error.message : 'Failed to fetch subscriptions',
        isLoadingSubscriptions: false
      });
    }
  },

  fetchMessages: async (eventTypes?: string[]) => {
    set({ isLoadingMessages: true, error: null });
    try {
      let endpoint = '/events/messages';
      if (eventTypes && eventTypes.length > 0) {
        const params = new URLSearchParams();
        eventTypes.forEach((type) => params.append('event_types', type));
        endpoint += `?${params.toString()}`;
      }
      const response = await apiClient.get(endpoint);
      if (!response.ok) {
        throw new Error(`Failed to fetch messages: ${response.status}`);
      }
      const messages = await response.json();
      set({ messages, isLoadingMessages: false });
    } catch (error) {
      set({
        error: error instanceof Error ? error.message : 'Failed to fetch messages',
        isLoadingMessages: false
      });
    }
  },

  fetchSubscription: async (subscriptionId: string) => {
    const response = await apiClient.get(`/events/subscriptions/${subscriptionId}`);
    if (!response.ok) {
      throw new Error(`Failed to fetch subscription: ${response.status}`);
    }
    return response.json();
  },

  createSubscription: async (request: CreateSubscriptionRequest) => {
    const response = await apiClient.post('/events/subscriptions', request);
    if (!response.ok) {
      throw new Error(`Failed to create subscription: ${response.status}`);
    }
    const subscription = await response.json();
    // Refresh subscriptions list
    get().fetchSubscriptions();
    return subscription;
  },

  updateSubscription: async (subscriptionId: string, request: UpdateSubscriptionRequest) => {
    const response = await apiClient.patch(`/events/subscriptions/${subscriptionId}`, request);
    if (!response.ok) {
      throw new Error(`Failed to update subscription: ${response.status}`);
    }
    const subscription = await response.json();
    // Refresh subscriptions list
    get().fetchSubscriptions();
    return subscription;
  },

  deleteSubscription: async (subscriptionId: string) => {
    const response = await apiClient.delete(`/events/subscriptions/${subscriptionId}`);
    if (!response.ok) {
      throw new Error(`Failed to delete subscription: ${response.status}`);
    }
    // Refresh subscriptions list
    get().fetchSubscriptions();
  },

  deleteSubscriptions: async (subscriptionIds: string[]) => {
    // Delete all subscriptions in parallel
    const results = await Promise.allSettled(
      subscriptionIds.map(id =>
        apiClient.delete(`/events/subscriptions/${id}`)
      )
    );

    // Check if any failed
    const failed = results.filter(r => r.status === 'rejected');
    if (failed.length > 0) {
      console.error(`Failed to delete ${failed.length} subscription(s)`);
    }

    // Refresh subscriptions list
    get().fetchSubscriptions();
  },

  fetchSubscriptionSecret: async (subscriptionId: string) => {
    const response = await apiClient.get(`/events/subscriptions/${subscriptionId}/secret`);
    if (!response.ok) {
      throw new Error(`Failed to fetch subscription secret: ${response.status}`);
    }
    return response.json();
  },

  clearEvents: () => {
    set({ subscriptions: [], messages: [], error: null });
  },
}));

File: backend/airweave/api/v1/endpoints/events.py
"""Events API endpoints for webhook subscriptions and event messages.

This module provides endpoints for managing webhook subscriptions and
retrieving event messages sent to those webhooks.
"""

from typing import List

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, HttpUrl
from svix.api import EndpointOut, EndpointSecretOut, MessageAttemptOut, MessageOut

from airweave.analytics import business_events
from airweave.api import deps
from airweave.api.context import ApiContext
from airweave.webhooks.constants.event_types import EventType
from airweave.webhooks.service import service as webhooks_service

router = APIRouter()


class SubscriptionWithAttemptsOut(BaseModel):
    """Response model for a subscription with its message attempts."""

    endpoint: EndpointOut
    message_attempts: List[MessageAttemptOut]

    class Config:
        """Pydantic config for arbitrary types."""

        arbitrary_types_allowed = True


class CreateSubscriptionRequest(BaseModel):
    """Request model for creating a new webhook subscription."""

    url: HttpUrl
    event_types: List[EventType]
    secret: str | None = None


class PatchSubscriptionRequest(BaseModel):
    """Request model for updating an existing webhook subscription."""

    url: HttpUrl | None = None
    event_types: List[EventType] | None = None


@router.get("/messages", response_model=List[MessageOut])
async def get_messages(
    ctx: ApiContext = Depends(deps.get_context),
    event_types: List[str] | None = Query(default=None),
) -> List[MessageOut]:
    """Get event messages for the current organization.

    Args:
        ctx: The API context containing organization info.
        event_types: Optional list of event types to filter by.

    Returns:
        List of event messages.
    """
    messages, error = await webhooks_service.get_messages(ctx.organization, event_types=event_types)
    if error:
        raise HTTPException(status_code=500, detail=error.message)
    return messages


@router.get("/messages/{message_id}", response_model=MessageOut)
async def get_message(
    message_id: str,
    ctx: ApiContext = Depends(deps.get_context),
) -> MessageOut:
    """Get a specific event message by ID.

    Args:
        message_id: The ID of the message to retrieve.
        ctx: The API context containing organization info.

    Returns:
        The event message with its payload.
    """
    message, error = await webhooks_service.get_message(ctx.organization, message_id)
    if error:
        raise HTTPException(status_code=500, detail=error.message)
    return message


@router.get("/messages/{message_id}/attempts", response_model=List[MessageAttemptOut])
async def get_message_attempts(
    message_id: str,
    ctx: ApiContext = Depends(deps.get_context),
) -> List[MessageAttemptOut]:
    """Get delivery attempts for a specific message.

    Args:
        message_id: The ID of the message.
        ctx: The API context containing organization info.

    Returns:
        List of delivery attempts for this message.
    """
    attempts, error = await webhooks_service.get_message_attempts_by_message(
        ctx.organization, message_id
    )
    if error:
        raise HTTPException(status_code=500, detail=error.message)
    return attempts or []


@router.get("/subscriptions", response_model=List[EndpointOut])
async def get_subscriptions(
    ctx: ApiContext = Depends(deps.get_context),
) -> List[EndpointOut]:
    """Get all webhook subscriptions for the current organization.

    Args:
        ctx: The API context containing organization info.

    Returns:
        List of webhook subscriptions.
    """
    endpoints, error = await webhooks_service.get_endpoints(ctx.organization)
    if error:
        raise HTTPException(status_code=500, detail=error.message)
    return endpoints


@router.get("/subscriptions/{subscription_id}", response_model=SubscriptionWithAttemptsOut)
async def get_subscription(
    subscription_id: str,
    ctx: ApiContext = Depends(deps.get_context),
) -> SubscriptionWithAttemptsOut:
    """Get a specific webhook subscription with its delivery attempts.

    Args:
        subscription_id: The ID of the subscription to retrieve.
        ctx: The API context containing organization info.

    Returns:
        The subscription details with message delivery attempts.
    """
    endpoint, error = await webhooks_service.get_endpoint(ctx.organization, subscription_id)
    if error:
        raise HTTPException(status_code=500, detail=error.message)

    message_attempts, attempts_error = await webhooks_service.get_message_attempts_by_endpoint(
        ctx.organization, subscription_id
    )
    if attempts_error:
        raise HTTPException(status_code=500, detail=attempts_error.message)

    return SubscriptionWithAttemptsOut(endpoint=endpoint, message_attempts=message_attempts or [])


@router.post("/subscriptions", response_model=EndpointOut)
async def create_subscription(
    request: CreateSubscriptionRequest,
    ctx: ApiContext = Depends(deps.get_context),
) -> EndpointOut:
    """Create a new webhook subscription.

    Args:
        request: The subscription creation request.
        ctx: The API context containing organization info.

    Returns:
        The created subscription.
    """
    endpoint, error = await webhooks_service.create_endpoint(
        ctx.organization, str(request.url), request.event_types, request.secret
    )
    if error:
        raise HTTPException(status_code=500, detail=error.message)

    # Track webhook subscription creation
    business_events.track_webhook_subscription_created(
        ctx=ctx,
        endpoint_id=endpoint.id,
        url=str(request.url),
        event_types=[e.value for e in request.event_types],
    )

    return endpoint


@router.delete("/subscriptions/{subscription_id}")
async def delete_subscription(
    subscription_id: str,
    ctx: ApiContext = Depends(deps.get_context),
) -> None:
    """Delete a webhook subscription.

    Args:
        subscription_id: The ID of the subscription to delete.
        ctx: The API context containing organization info.
    """
    error = await webhooks_service.delete_endpoint(ctx.organization, subscription_id)
    if error:
        raise HTTPException(status_code=500, detail=error.message)

    # Track webhook subscription deletion
    business_events.track_webhook_subscription_deleted(ctx=ctx, endpoint_id=subscription_id)


@router.patch("/subscriptions/{subscription_id}", response_model=EndpointOut)
async def patch_subscription(
    subscription_id: str,
    request: PatchSubscriptionRequest,
    ctx: ApiContext = Depends(deps.get_context),
) -> EndpointOut:
    """Update a webhook subscription.

    Args:
        subscription_id: The ID of the subscription to update.
        request: The subscription update request.
        ctx: The API context containing organization info.

    Returns:
        The updated subscription.
    """
    url = str(request.url) if request.url else None
    endpoint, error = await webhooks_service.patch_endpoint(
        ctx.organization, subscription_id, url, request.event_types
    )
    if error:
        raise HTTPException(status_code=500, detail=error.message)

    # Track webhook subscription update
    business_events.track_webhook_subscription_updated(
        ctx=ctx,
        endpoint_id=subscription_id,
        url_changed=request.url is not None,
        event_types_changed=request.event_types is not None,
        new_event_types=[e.value for e in request.event_types] if request.event_types else None,
    )

    return endpoint


@router.get("/subscriptions/{subscription_id}/secret", response_model=EndpointSecretOut)
async def get_subscription_secret(
    subscription_id: str,
    ctx: ApiContext = Depends(deps.get_context),
) -> EndpointSecretOut:
    """Get the signing secret for a webhook subscription.

    Args:
        subscription_id: The ID of the subscription.
        ctx: The API context containing organization info.

    Returns:
        The subscription's signing secret.
    """
    secret, error = await webhooks_service.get_endpoint_secret(ctx.organization, subscription_id)
    if error:
        raise HTTPException(status_code=500, detail=error.message)

    # Track webhook secret viewing (security audit trail)
    business_events.track_webhook_secret_viewed(ctx=ctx, endpoint_id=subscription_id)

    return secret
